libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Generic.cpp

```diff
@@ -107,60 +106,37 @@ void AP_BattMonitor_SMBus_Generic::timer()
         _state.last_time_micros = tnow;
     }
 
-    read_full_charge_capacity();
-
     // FIXME: Perform current integration if the remaining capacity can't be requested
     read_remaining_capacity();
 
     read_temp();
-
-    read_serial_number();
-
-    read_cycle_count();
 }
 
-// read_block - returns number of characters read if successful, zero if unsuccessful
-uint8_t AP_BattMonitor_SMBus_Generic::read_block(uint8_t reg, uint8_t* data, bool append_zero) const
+// Read batery information messages once: PEC, cycles, serial number, device name, manufacturer name
+bool AP_BattMonitor_SMBus_Generic::initialize()
 {
-    // get length
-    uint8_t bufflen;
-    // read byte (first byte indicates the number of bytes in the block)
-    if (!_dev->read_registers(reg, &bufflen, 1)) {
-        return 0;
-    }
-
-    // sanity check length returned by smbus
-    if (bufflen == 0 || bufflen > SMBUS_READ_BLOCK_MAXIMUM_TRANSFER) {
-        return 0;
-    }
-
-    // buffer to hold results (2 extra byte returned holding length and PEC)
-    const uint8_t read_size = bufflen + 1 + (_pec_supported ? 1 : 0);
-    uint8_t buff[read_size];
-
-    // read bytes
-    if (!_dev->read_registers(reg, buff, read_size)) {
-        return 0;
-    }
-
-    // check PEC
-    if (_pec_supported) {
-        uint8_t pec = get_PEC(AP_BATTMONITOR_SMBUS_I2C_ADDR, reg, true, buff, bufflen+1);
-        if (pec != buff[bufflen+1]) {
-            return 0;
-        }
+    if(_initialized){
+        return true;
     }
-
-    // copy data (excluding PEC)
-    memcpy(data, &buff[1], bufflen);
-
-    // optionally add zero to end
-    if (append_zero) {
-        data[bufflen] = '\0';
+    
+    read_full_charge_capacity();
+    read_serial_number();
+    read_cycle_count();
+    read_device_name();
+    read_manufacturer_name();
+    read_design_capacity();
+
+    // Set parameters with retrieved data
+    read();
+    
+    // check if PEC is supported
+    if (!check_pec_support()) {
```

Comment:
PEC checks should always be done before any thing else. In fact until you've confirmed PEC either way you should not be attempting to read any other data from the battery. If we have error checking we need it to be used. That should be done before reading any other stats from the battery. There is a reason we always return if we haven't checked the PEC support in what was already here.

Reply:
Sorry about that bug.  I got it in the wrong position while trying to limit how often all of the system_info functions got called.